### PR TITLE
Load Balancing implementation + other fixes

### PR DESCRIPTION
**Pull request by bltnynk**
WIP
## Stuff to do
- [x] Implement token-wise load balancing
- [x] Adapt load balancing to sequence-wise (since we are routing by sequence).
- [ ] Fix possible bugs in z-loss and load balancing loss.
- [ ] Other stuff.

### CODE REVIEW

File: src/levanter/main/routed_lm.py [hunk 1]
```diff
@@ -71,6 +74,35 @@ def __post_init__(self):
...
+    if model.config.sequence_load_balancing:
+        expert_mask /= seq_length
+
+    # Compute the percentage of tokens routed to each experts    
+    tokens_per_expert = hax.mean(expert_mask, axis=[Batch, Pos], where=router_hs_idxs>=0)
+
+    # Compute the average probability of routing to these experts
+    router_prob_per_expert = expert_latest_load
+
+    lb_loss = hax.sum(tokens_per_expert * router_prob_per_expert) * Experts.size / TopK.size
```
Comment (Aphoh) [inline src/levanter/main/routed_lm.py]: Should be `(Experts.size -TopK.size)/TopK.size`

Comment (Aphoh) [inline src/levanter/main/routed_lm.py]: We need to keep the loss per batch element, so just summing over `[Pos]` here instead of `Batch` as well. Otherwise microbatching gets broken... 😢

File: src/levanter/models/routed_qwen_model.py [hunk 2]
```diff
@@ -732,11 +737,25 @@ def __call__(
 
 
 # Modified LM head model for Qwen
-class Router(hnn.Linear):
+class Router(eqx.Module):
+    linear: hnn.Linear
+    expert_bias: NamedArray
+    expert_mean_load: RunningMean
```
Comment (Aphoh) [inline src/levanter/models/routed_qwen_model.py]: These need to get ignored in the `is_trainable` mask... I think?

File: src/levanter/models/routed_qwen_model.py [hunk 3]
```diff
@@ -910,6 +936,15 @@ def routed_forward(
         else:
             expert_mask = hax.where(router_hs_idxs < 0, 0.0, expert_mask).astype(compute_dtype)
 
+        # Compute the number of tokens routed to each experts, then update the expert bias.
+        used_experts = expert_mask > 0
+        if self.config.sequence_load_balancing:
+            used_experts /= seq_length
+        expert_latest_load = hax.sum(used_experts, (Batch, Pos), where=router_hs_idxs>=0, dtype=compute_dtype)
+        self.router.expert_mean_load.add(expert_latest_load, 1)
```
Comment (Aphoh) [inline src/levanter/models/routed_qwen_model.py]: This is immutable here

File: src/levanter/models/routed_qwen_model.py [hunk 4]
```diff
@@ -910,6 +936,15 @@ def routed_forward(
...
+        expert_bias_update = self.config.expert_bias_update_rate * hax.where(expert_latest_load < self.router.expert_mean_load.mean, 1, -1)
+        dataclasses.replace(self.router, expert_bias=self.router.expert_bias + expert_bias_update)
```
Comment (Aphoh) [inline src/levanter/models/routed_qwen_model.py]: This also does nothing... `dataclasses.replace` returns a new copy of `self.router` here.

**Review by Aphoh — CHANGES_REQUESTED** [review]
* just get aux loss stuff working here

File: src/levanter/main/routed_lm.py (hunk 1 above)
Comment (bltnynk) [reply src/levanter/main/routed_lm.py]: Fixed

File: src/levanter/models/routed_qwen_model.py (hunk 3 above)
Comment (bltnynk) [reply src/levanter/models/routed_qwen_model.py]: Fixed

File: src/levanter/models/routed_qwen_model.py (hunk 4 above)
Comment (bltnynk) [reply src/levanter/models/routed_qwen_model.py]: Fixed

File: src/levanter/models/routed_qwen_model.py (hunk 2 above)
Comment (bltnynk) [reply src/levanter/models/routed_qwen_model.py]: Done
